packages/tooling/fast-tooling/src/data-utilities/shortcut-action.ts

```diff
@@ -0,0 +1,107 @@
+export interface MetaKey {
+    metaKey: true;
+}
+
+export interface ShiftKey {
+    shiftKey: true;
+}
+
+export interface CtrlKey {
+    ctrlKey: true;
+}
+
+export interface AltKey {
+    altKey: true;
+}
+
+export type ModifierKey = AltKey | CtrlKey | ShiftKey | MetaKey;
+
+export interface SpecificKey {
+    /**
+     * The key string value
+     */
+    value: string;
+}
+
+export type KeyConfig = SpecificKey | ModifierKey;
+
+export interface ShortcutActionConfig {
+    /**
+     * The display name of the shortcut action
+     */
+    name: string;
+
+    /**
+     * The keys needed to execute the action
+     */
+    keys: KeyConfig[];
+
+    /**
+     * The action to take when the keycodes have been pressed
+     */
+    action: () => void;
+}
+
+export class ShortcutAction {
+    private action: () => void;
+    public keys: KeyConfig[];
+    public name: string;
+
+    constructor(config: ShortcutActionConfig) {
+        this.action = config.action;
+        this.keys = config.keys;
+        this.name = config.name;
+    }
+
+    public runAction = (keys: KeyConfig[]): void => {
+        if (this.shouldRunAction(keys)) {
+            this.action();
+        }
+    };
+
+    private shouldRunAction = (keys: KeyConfig[]): boolean => {
```

Comment:
Practically, this will almost always be run in response to a keyboard event. Per my comment on the spec, I'd expect an API that took that event object as input. If I'm reading this right, with this API, we'd have to have an additional translation layer to translate the keyboard event into an array of KeyConfig objects before this API could be called.

As a consumer, I think I'd like to have something like this:

```js
if (action.matches(event)) {
  action.invoke();
}
```